code/modules/reagents/withdrawl/_addiction.dm

```diff
@@ -0,0 +1,136 @@
+/datum/addiction/proc/on_lose_addiction_points(datum/mind/victim_mind)
+	var/current_addiction_point_amount = victim_mind.addiction_points[type]
+	if(!LAZYACCESS(victim_mind.active_addictions, type)) //Not addicted
+		return FALSE
+	if(current_addiction_point_amount > addiction_loss_threshold) //Not enough to stop being addicted
+		return FALSE
+	lose_addiction(victim_mind)
+	return TRUE
+
+/datum/addiction/proc/lose_addiction(datum/mind/victim_mind)
+	SEND_SIGNAL(victim_mind.current, COMSIG_CLEAR_MOOD_EVENT, "[type]_addiction")
+	SEND_SIGNAL(victim_mind.current, COMSIG_CARBON_LOSE_ADDICTION, victim_mind)
+	to_chat(victim_mind.current, span_notice("You feel like you've gotten over your need for drugs."))
+	end_withdrawal(victim_mind.current)
+	LAZYREMOVE(victim_mind.active_addictions, type)
+
+/datum/addiction/proc/process_addiction(mob/living/carbon/affected_carbon, delta_time, times_fired)
+	var/current_addiction_cycle = LAZYACCESS(affected_carbon.mind.active_addictions, type) //If this is null, we're not addicted
+	var/on_drug_of_this_addiction = FALSE
+	for(var/datum/reagent/possible_drug as anything in affected_carbon.reagents.reagent_list) //Go through the drugs in our system
+		for(var/addiction in possible_drug.addiction_types) //And check all of their addiction types
+			if(addiction == type && possible_drug.volume >= addiction_relief_treshold) //If one of them matches, and we have enough of it in our system, we're not losing addiction
+				if(current_addiction_cycle)
+					end_withdrawal(affected_carbon) //stop the pain
+				on_drug_of_this_addiction = TRUE
+				break
+
+	var/withdrawal_stage
+
+	switch(current_addiction_cycle)
+		if(WITHDRAWAL_STAGE1_START_CYCLE to WITHDRAWAL_STAGE1_END_CYCLE)
+			withdrawal_stage = 1
+		if(WITHDRAWAL_STAGE2_START_CYCLE to WITHDRAWAL_STAGE2_END_CYCLE)
+			withdrawal_stage = 2
+		if(WITHDRAWAL_STAGE3_START_CYCLE to INFINITY)
+			withdrawal_stage = 3
+		else
+			withdrawal_stage = 0
+
+	if(!on_drug_of_this_addiction)
+		if(affected_carbon.mind.remove_addiction_points(type, addiction_loss_per_stage[withdrawal_stage + 1] * delta_time)) //If true was returned, we lost the addiction!
+			return
+
+	if(!current_addiction_cycle) //Dont do the effects if were not on drugs
+		return FALSE
+
+	switch(current_addiction_cycle)
+		if(WITHDRAWAL_STAGE1_START_CYCLE)
+			withdrawal_enters_stage_1(affected_carbon)
+		if(WITHDRAWAL_STAGE2_START_CYCLE)
+			withdrawal_enters_stage_2(affected_carbon)
+		if(WITHDRAWAL_STAGE3_START_CYCLE)
+			withdrawal_enters_stage_3(affected_carbon)
+
+	///One cycle is 2 seconds
+	switch(withdrawal_stage)
+		if(1)
+			withdrawal_stage_1_process(affected_carbon, delta_time)
+		if(2)
+			withdrawal_stage_2_process(affected_carbon, delta_time)
+		if(3)
+			withdrawal_stage_3_process(affected_carbon, delta_time)
+
+	LAZYADDASSOC(affected_carbon.mind.active_addictions, type, 1 * delta_time) //Next cycle!
+
+/// Called when addiction enters stage 1
```

Comment:
These stages are very similar to disease stages.
Could probably point this out in the changelog for future coders, can put interesting effects with new chems.